### ISSUE TITLE
Imitation Learning Data processing & Training codebase

### ISSUE THREAD

**Issue by adityakotha03** [Veda-Ashok/cooking_up_collaboration#6]
This update is inspired by this [repo](https://github.com/bic4907/Overcooked-AI/tree/main) and the Overcooked AI repo. Currently, Overcooked AI doesn't support BC or RL because TensorFlow's dependencies are deprecated and not well-maintained. So this patch uses Torch to do BC instead. This currently supports LSTM and MLP models for BC. It also supports the webview application to use the trained PyTorch models.

**Comment by adityakotha03** [comment]
## Behaviour Cloning

You can download the original data collected in the overcooked AI repo [here](https://drive.google.com/drive/folders/1aGV8eqWeOG5BMFdUcVoP2NHU_GFPqi57) and then palce it under `data/` in the root folder of this repo.

Right now BC training is setup from csv first (`data/2019_hh_trials.csv`) and supports both models:
- LSTM (default)
- MLP

To train:
```bash
python train_bc.py --model lstm --run-name bc_lstm_v1
```
or
```bash
python train_bc.py --model mlp --run-name bc_mlp_v1
```

Default training uses both player perspectives in one shared policy (`--player-mode both`).

To train single-player BC for ablations:
```bash
python train_bc.py --model lstm --player-mode single --player-idx 0 --run-name bc_lstm_p0
```

### What data is actually used as input?

For each timestep row in csv:
- `state` is parsed into OvercookedState
- then we run Overcooked featurizer (`mdp.featurize_state`)
- in default mode (`--player-mode both`), both player 0 and player 1 views are added as supervised samples
- this gives feature vectors (typically 96 dim)

Label/target is from `joint_action`:
- default: both players are used (shared policy training)
- optional ablation: only one selected player with `--player-mode single --player-idx {0|1}`

Action space is 6 classes:
- `UP`, `DOWN`, `LEFT`, `RIGHT`, `STAY`, `INTERACT`

### Are we training all layouts in csv?

Yes. Current pipeline uses all layouts available in the csv and then does split by `trial_id` (not by layout holdout).  
So if your csv has multiple layouts (like `cramped_room`, `coordination_ring`, `asymmetric_advantages`, `random0`, `random3`), all of them are used.

### Training outputs

Each run saves files like:
- `best.pt`
- `last.pt`
- `metrics.json`
- `config.json`
- `split_summary.json`
- `preprocessing_report.json`

Default output dir right now is:
```bash
trained_models/bc/
```

## Running BC Models in the Webapp (PyTorch)

Now webapp can load BC PyTorch models directly (no tensorflow conversion needed).

It works by adding an agent folder under:
```bash
webapp/server/static/assets/agents/<YourAgentName>/
```

Put these files inside:
- `best.pt`
- `agent_manifest.json`

Then restart webapp:
```bash
cd webapp
./up.sh
```

Your folder name shows up in agent dropdown automatically.

### Example manifest (MLP)

```json
{
  "type": "bc_torch",
  "model_type": "mlp",
  "checkpoint": "best.pt",
  "supported_layouts": ["cramped_room", "coordination_ring", "asymmetric_advantages", "random0", "random3"],
  "input_dim": 96,
  "num_actions": 6,
  "mlp_hidden": [256, 128],
  "dropout": 0.1,
  "planner_cache_dir": ".cache/overcooked_planners"
}
```

### Example manifest (LSTM)

```json
{
  "type": "bc_torch",
  "model_type": "lstm",
  "checkpoint": "best.pt",
  "supported_layouts": ["cramped_room", "coordination_ring", "asymmetric_advantages", "random0", "random3"],
  "input_dim": 96,
  "num_actions": 6,
  "seq_len": 20,
  "hidden_dim": 128,
  "num_layers": 1,
  "dropout": 0.1,
  "planner_cache_dir": ".cache/overcooked_planners"
}
```

Note:
- If selected layout is not in `supported_layouts`, BC agent safely returns `STAY`.
- `player_idx` in manifest is optional. If omitted, the agent uses its runtime slot (player 0 or player 1).
- Since server requirements now include `torch`, do a docker rebuild if needed.

**Comment by adityakotha03** [comment]
Also, the Docker up.sh commands are changed now

By default, `./up.sh` reuses the existing Docker image (no forced rebuild).  
By default, it runs in the foreground (you see live logs).  
Optional flags:
```bash
./up.sh --build           # build with cache, then run
./up.sh --rebuild         # no-cache rebuild, force recreate
./up.sh --detach          # run in background
./up.sh production --build
./up.sh production --rebuild
./up.sh production --detach
```

**Comment by adityakotha03** [comment]
Also, a new update adds a Sampling strategy to the existing trained models, because the deterministic argmax keeps it from moving, so adding randomness helps diversify the possible moves.

```json
{
  "type": "bc_torch",
  "model_type": "mlp",
  "checkpoint": "best.pt",
  "sampling_mode": "sample",
  "sampling_temperature": 1.0,
  "supported_layouts": ["cramped_room", "coordination_ring", "asymmetric_advantages", "random0", "random3"],
  "input_dim": 96,
  "num_actions": 6,
  "mlp_hidden": [256, 128],
  "dropout": 0.1,
  "planner_cache_dir": ".cache/overcooked_planners"
}
```

It would look something like this, with sampling_mode and temperature added. Options are between argmax, which is deterministic, and sample.
